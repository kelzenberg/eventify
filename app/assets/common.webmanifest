{
    "icons": [
        { "src": "/assets/Favicon_192.png", "type": "image/png", "sizes": "192x192" },
        { "src": "/assets/Favicon_512.png", "type": "image/png", "sizes": "512x512" }
    ]
}