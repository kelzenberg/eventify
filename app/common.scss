@use "sass:map";
@import "./bootstrap.custom.scss";

body {
    margin: 0px;
    font-family: 'Athiti', sans-serif;
}

.fancy-background {
    position: fixed;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
    z-index: -5;
    overflow: hidden;
    background: linear-gradient(160deg, #F6CAA2, #F2B680);

    &::before {
        content: "";
        position: absolute;
        z-index: 0;
        top: 0px;
        left: 0px;
        width: 200vw;
        height: 30vh;
        transform: rotateZ(-20deg);
        transform-origin: 0% 100%;
        background: #f5ceaa;
        filter: blur(5px);
    }

    &::after {
        content: "";
        position: absolute;
        z-index: 0;
        bottom: 0px;
        right: 0px;
        width: 200vw;
        height: 30vh;
        transform: rotateZ(-20deg);
        transform-origin: 100% 0%;
        background: #F6CAA2;
        filter: blur(5px);
    }
}

.iconBox {
    display: inline-block;
    position: relative;
    border-radius: 13px;
    border: 1px solid map.get($theme-colors, "primary");
    width: 50px;
    height: 50px;
    background-repeat: no-repeat;
    background-position: 50% 50%;
    vertical-align: middle;

    img {
        width: 100%;
        height: 100%;
    }

    @each $color, $value in $theme-colors {
        &.iconBox-#{$color} {
            border-color: $value;
        }
    }

    &.iconBox-removable:hover::after {
        content: "";
        position: absolute;
        top: 0px;
        left: 0px;
        right: 0px;
        bottom: 0px;
        border-radius: 13px;
        background-color: rgba(255, 255, 255, 0.5);
        background-image: url("/assets/icons/remove.svg");
        background-repeat: no-repeat;
        background-position: 50% 50%;
        pointer-events: none;
    }
}

.hover-accent:hover {
    background-color: #ececec;
}