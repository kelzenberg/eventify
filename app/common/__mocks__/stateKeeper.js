const stateKeeper = jest.requireActual("../stateKeeper");

export const UserContext = stateKeeper.UserContext;
export const isAuthenticated = jest.fn();
export const setCredentials = jest.fn();
export const getAuthenticationToken = jest.fn();
export const getUserInfo = jest.fn();
export const maybeUpdateUserInfo = jest.fn();
export const clearLogin = jest.fn();