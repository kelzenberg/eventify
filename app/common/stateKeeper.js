import React from 'react';

export const UserContext = React.createContext(null);

export function isAuthenticated() {
    return localStorage.getItem("USER_TOKEN") != null;
}

export function getDisplayName() {
    return "Username";
}

export function setCredentials(token, userInfo) {
    localStorage.setItem("USER_TOKEN", token);
}

export function getAuthenticationToken() {
    return localStorage.getItem("USER_TOKEN");
}

// getUserInfo returns the stored user information or dummy data if none is set
export function getUserInfo() {
    try {
        let data = JSON.parse(localStorage.getItem("USER_INFO"));
        if (data == null) throw new Error();
        return data;
    } catch(e) {
        return {
            "id": "",
            "createdAt": "2021-06-02T00:06:45.528+00:00",
            "updatedAt": "2021-06-02T00:06:45.528+00:00",
            "email": "",
            "displayName": "",
            "authRole": ""
        };
    }
}

// maybeUpdateUserInfo takes a user info structure and checks if it matches the stored information.
// If the informations match it does nothing and returns false.
// If the new user information differs from the stored one it saves the new data and returns true.
export function maybeUpdateUserInfo(newUserInfo) {
    if(localStorage.getItem("USER_INFO") === null) {
        setUserInfo(newUserInfo);
        return true;
    }
    let oldUserInfo = getUserInfo();
    for(let key in newUserInfo) {
        if(newUserInfo[key] !== oldUserInfo[key]) {
            setUserInfo(newUserInfo);
            return true;
        }
    }
    return false;
}

function setUserInfo(userInfo) {
    localStorage.setItem("USER_INFO", JSON.stringify(userInfo));
}

export function clearLogin() {
    localStorage.removeItem('USER_TOKEN');
    localStorage.removeItem('USER_INFO');
}