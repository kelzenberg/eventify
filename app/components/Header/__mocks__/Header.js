import React from "react";

export default jest.fn(() => {return <div>Mocked Header Component</div>});