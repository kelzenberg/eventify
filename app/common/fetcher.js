import * as stateKeeper from "./stateKeeper";

const apiHost = "https://localhost:8443/";

class FetchError extends Error {
    constructor(msg, code) {
        super(msg);
        this.code = code;
    }
}

function request(options) {
    if(typeof options.body == "object") {
        options.body = JSON.stringify(options.body);
    }
    var additionalHeaders = {};
    if(options.headers !== undefined) {
        additionalHeaders = options.headers
    }
    if(options.parameters != undefined) {
        options.path += "?" + new URLSearchParams(options.parameters).toString();
    }
    let authorization = null;
    if(stateKeeper.isAuthenticated()) {
        authorization = `Bearer ${stateKeeper.getAuthenticationToken()}`;
    }
    return new Promise(function(resolve, reject){
        fetch(apiHost + options.path, {
            credentials: 'omit',
            body: options.body,
            headers: {
                'accept': 'application/json',
                'content-type': 'application/json',
                'authorization': authorization,
                ...additionalHeaders
            },
            method: options.method,
            mode: 'cors',
            redirect: 'follow',
            referrer: 'no-referrer',
            keepalive: options.keepalive === true
        })
        // Check for errors
        .then(response => {
            if(response.ok || (options.allowedStatuses != undefined && options.allowedStatuses.includes(response.status)))Â {
                return Promise.resolve(response)
            } else if(response.status >= 400) {
                return Promise.reject(new FetchError(`request returned with status code ${response.status}`, response.status));
            }
        })
        // Decode content
        .then(response => {
            var contentType = response.headers.get("content-type");
            if(contentType == null) {
                if(options.expect != undefined && options.expect != "empty") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return Promise.resolve([null, response]);
            }
            if(contentType.includes("application/json")) {
                if(options.expect != undefined && options.expect != "json") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return response.json().then(data => {
                    return [data, response];
                })
            } else if(contentType.includes("text/html")) {
                if(options.expect != undefined && options.expect != "text") {
                    console.warn(`fetcher: got response with content type '${contentType}', but expected '${options.expect}'.`);
                    return Promise.reject(new FetchError(`Invalid content-type '${contentType}', but expected '${options.expect}'`));
                }
                return response.text().then(data => {
                    return [data, response];
                })
            }
        })
        // Parse data
        .then(([data, response]) => {
            resolve([data, response]);
        })
        // Handle remaining errors
        .catch(error => {
            if(error instanceof FetchError)
                reject(error);
            else {
                console.log(error);
                reject(new FetchError("fetcher encountered an unexpected error: " + error.message, -100))
            }
        });
    });
}

const status = {
    continue: 100,
    switchingProtocols: 101,
    processing: 102,
    earlyHints: 103,
    ok: 200,
    created: 201,
    accepted: 202,
    nonAuthorativeInformation: 203,
    noContent: 204,
    resetContent: 205,
    partialContent: 206,
    alreadyReported: 208,
    imUsed: 226,
    multipleChoice: 300,
    movedPermanently: 301,
    found: 302,
    seeOther: 303,
    notModified: 304,
    useProxy: 305,
    temporaryRedirect: 307,
    permanentRedirect: 308,
    badRequest: 400,
    unauthorized: 401,
    paymentRequired: 402,
    forbidden: 403,
    notFound: 404,
    methodNotAllowed: 405,
    notAcceptable: 406,
    proxyAuthenticationRequired: 407,
    requestTimeout: 408,
    conflict: 409,
    gone: 410,
    lengthRequired: 411,
    preconditionFailed: 412,
    payloadTooLarge: 413,
    uriTooLong: 414,
    unsupportedMediaType: 415,
    requestedRangeNotSatisfiable: 416,
    expectationFailed: 417,
    imATeapot: 418,
    misdirectedRequest: 421,
    unprocessedEntity: 422,
    locked: 423,
    failedDependency: 424,
    tooEarly: 425,
    upgradeRequired: 426,
    preconditionRequired: 428,
    tooManyRequests: 429,
    requestHeaderFieldsTooLarge: 431,
    unavailableForLegalReasons: 451,
    internalServerError: 500,
    notImplemented: 501,
    badGateway: 502,
    serviceUnavailable: 503,
    gatewayTimeout: 504,
    httpVersionNotSupported: 505,
    variantAlsoNegotiates: 506,
    insufficientStorage: 507,
    loopDetected: 508,
    notExtended: 510,
    networkAuthenticationRequired: 511
};

export default {
    FetchError,
    request,
    status
}