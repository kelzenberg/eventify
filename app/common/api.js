import fetcher from "./fetcher";
import * as stateKeeper from "./stateKeeper";

export class UnexpectedServerError extends Error {
    constructor(msg, code) {
        super(msg);
    }
}

export function authenticate(email, password) {
    return new Promise((resolve, reject) => {
        fetcher.request({
            method: "POST",
            path: "authenticate",
            expect: "json",
            body: {"email": email, "password": password},
            allowedStatuses: [fetcher.status.unauthorized]
        })
        .then(([responseData, rawResponse]) => {
            if(rawResponse.status === fetcher.status.unauthorized) {
                reject();
            } else {
                stateKeeper.setCredentials(responseData.token);
                resolve();
            }
        })
        .catch(err => {
            reject(new UnexpectedServerError(err));
        })
    });
}

export function getUserInfo() {
    return fetcher.request({method: "GET", path: "me", expect: "json"}).then(([response]) => response);
}

export function register(email, password, displayName) {
    let body = {
        "email": email,
        "password": password,
        "displayName": displayName
    };
    return fetcher.request({method: "POST", path: "register", expect: "json", body: body})
    .then(([responseData, rawResponse]) => {
        stateKeeper.setCredentials(responseData.token);
        return Promise.resolve();
    })
}

export function logout() {
    console.warn("TODO: api logout");
}

// =========================================== EVENTS =========================================== //

export function getAllUserEvents() {
    return fetcher.request({method: "GET", path: "me/events", expect: "json"}).then(([response]) => response);
}

export function getEvent(eventID) {
    return fetcher.request({method: "GET", path: `events/${eventID}`, expect: "json"}).then(([response]) => response);

    return Promise.resolve({
        "id": "22f7c93e-cfdc-491f-baa2-732fa298d8f6",
        "createdAt": "2021-06-02T15:42:16.478+00:00",
        "updatedAt": "2021-06-02T15:42:16.478+00:00",
        "title": "TestEvent 1",
        "description": "This is a test description",
        "startedAt": "2021-12-24T13:33:37.111+00:00",
        "endedAt": null,
        "expenseSharingModules": [
            {
                "id": "0f994c11-58ba-4718-ae18-5fe099177329",
                "createdAt": "2021-06-02T15:42:40.985+00:00",
                "updatedAt": "2021-06-02T15:42:40.985+00:00",
                "title": "Expense Sharing Module 1",
                "description": "This is a test description",
                "payments": [
                    {
                        "id": "e2ba702a-3713-4804-9b67-de5f9f99b3a8",
                        "createdAt": "2021-06-02T15:43:05.301+00:00",
                        "updatedAt": "2021-06-02T15:43:05.301+00:00",
                        "title": "Payment 1",
                        "amount": 13.37,
                        "payer": {
                            "id": "1798a949-1f72-45ff-b83c-d8e796add4cc",
                            "createdAt": "2021-05-29T17:01:31.716+00:00",
                            "updatedAt": "2021-05-29T17:01:31.716+00:00",
                            "email": "admin@test.de",
                            "displayName": "Admin",
                            "authRole": "ADMIN"
                        },
                        "shareType": "FIXED"
                    }
                ]
            }
        ]
    });
}

export function saveEvent(event) {
    return fetcher.request({method: "PUT", path: `events/${event.id}`, expect: "json"}).then(([response]) => response);
}