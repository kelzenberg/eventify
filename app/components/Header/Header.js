import React from 'react';
import * as api from '../../common/api';
import * as stateKeeper from '../../common/stateKeeper';

export default function Header() {
    let userInfo = React.useContext(stateKeeper.UserContext);

    return <div className="container-fluid p-4">
        <div className="d-flex">
            <a href="/you" className="me-auto">
                <img src="/assets/logo_text.svg" className="bi me-2" height="40" aria-label="Eventify"/>
            </a>
            <div className="col-md-3 text-end">
                <span className="text-muted">Logged in as {userInfo.email}</span>
                <button className="btn btn-outline-darkGray ms-3" onClick={logout}>Logout</button>
            </div>
        </div>
    </div>
}

function logout() {
    api.logout();
    stateKeeper.clearLogin();
    window.location = "/";
}