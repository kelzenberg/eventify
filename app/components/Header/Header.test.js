import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import { default as snapshotRenderer } from 'react-test-renderer';
import { UserContext } from "../../common/stateKeeper";
import Header from "../../components/Header/Header";
import * as stateKeeper from "../../common/stateKeeper";

const userInfo = {
    id: "123456789",
    displayName: "Jane Doe",
    email: "one@two.three"
}

test('Snapshot', () => {
    let component = snapshotRenderer.create(
        <UserContext.Provider value={userInfo}>
            <Header/>
        </UserContext.Provider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('logout', () => {
    const {queryByText, getByText} = render(
        <UserContext.Provider value={userInfo}>
            <Header/>
        </UserContext.Provider>
    );

    expect(queryByText(userInfo.displayName, {exact: false})).toBeTruthy();

    let clearLogin = jest.spyOn(stateKeeper, "clearLogin");
    fireEvent.click(getByText("Logout"));
    expect(clearLogin).toHaveBeenCalled();
    clearLogin.mockRestore();
})