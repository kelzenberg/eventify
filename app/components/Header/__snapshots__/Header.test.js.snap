// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<div
  className="container-fluid p-4"
>
  <div
    className="d-flex"
  >
    <a
      className="me-auto"
      href="/you"
    >
      <img
        aria-label="Eventify"
        className="bi me-2"
        height="40"
        src="/assets/logo_text.svg"
      />
    </a>
    <div
      className="col-md-3 text-end"
    >
      <span
        className="text-muted"
      >
        Logged in as 
        Jane Doe
      </span>
      <button
        className="btn btn-outline-darkGray ms-3"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</div>
`;
